libs/utils/src/filter.c

```diff
@@ -17,421 +17,497 @@
-    char *sub = calloc(strlen(filterString) + 1, sizeof(*sub));
-    celix_array_list_t* operands = celix_arrayList_create();
-    int keepRunning = 1;
+    int rc = fclose(celix_steal_ptr(stream));
+    if (rc != 0) {
+        celix_err_push("Filter Error: Failed to close stream.");
+        return NULL;
+    }
 
-    while (keepRunning) {
-        char c = filterString[*pos];
-        
+    return celix_steal_ptr(value);
+}
 
-        switch (c) {
-            case ')': {
-                if (strlen(sub) > 0) {
-                    celix_arrayList_add(operands, strdup(sub));
-                }
-                keepRunning = 0;
-                break;
-            }
-            case '\0':{
-                fprintf(stderr, "Filter Error: Unclosed bracket.\n");
-                keepRunning = false;
-                break;
-            }
-            case '(': {
-                fprintf(stderr, "Filter Error: Invalid value.\n");
-                keepRunning = false;
-                break;
-            }
-            case '*': {
-                if (strlen(sub) > 0) {
-                    celix_arrayList_add(operands, strdup(sub));
-                }
-                sub[0] = '\0';
-                celix_arrayList_add(operands, NULL);
-                (*pos)++;
-                break;
-            }
-            case '\\': {
-                (*pos)++;
-                c = filterString[*pos];
-            }
-            /* no break */
-            default: {
-                char ch[2];
-                ch[0] = c;
-                ch[1] = '\0';
-                strcat(sub, ch);
-                (*pos)++;
-                break;
+static bool celix_filter_isSubString(const char* filterString, int startPos) {
+    for (int i = startPos; filterString[i] != '\0' && filterString[i] != ')'; i++) {
+        if (filterString[i] == '*') {
+            return true;
+        }
+    }
+    return false;
+}
+
+static celix_status_t celix_filter_parseSubstringAny(const char* filterString, int* pos, char** out) {
+    celix_autofree char* any = NULL;
+    size_t anySize = 0;
+    celix_autoptr(FILE) stream = NULL;
+    while (filterString[*pos] != ')' && filterString[*pos] != '*') {
+        int rc;
+        if (filterString[*pos] == '\\') {
+            (*pos)++; // eat '\'
+        }
+        if (filterString[*pos] == '\0') {
+            celix_err_push("Filter Error: Unexpected end of string while parsing attribute value.");
+            return CELIX_INVALID_SYNTAX;
+        }
+        if (!stream) {
+            stream = open_memstream(&any, &anySize);
+            if (!stream) {
+                celix_err_push("Filter Error: Failed to open mem stream.");
+                return CELIX_ENOMEM;
             }
         }
+        rc = fputc(filterString[*pos], stream);
+        if (rc == EOF) {
+            celix_err_push("Filter Error: Failed to write to stream.\n");
+            return CELIX_FILE_IO_EXCEPTION;
```

Comment:
If `fputc` into a memstream fails, the most probably error might be `CELIX_ENOMEM` rather than `CELIX_FILE_IO_EXCEPTION`.